### PR TITLE
Add a guide to maintain nice pull requests to Contributing.md

### PR DESCRIPTION
**Pull request by jhass**
Feedback welcome

### CODE REVIEW

**Comment by jreinert** [comment]
LGTM

**Comment by vyp** [comment]
+1. Although it seems like you are also teaching people git rather than just giving pull request/commit guidelines (the first paragraph). No objection to a little git tutorial of course, but maybe separate into further subheadings? Like perhaps "Commit Guidelines" and "How to Rebase" or something?

Also I think "Other useful flags to `git commit` are `--fixup` and `--squash`, ..." reads better?

**Comment by jhass** [comment]
Updated the wording and added a subsubheader

**Comment by vyp** [comment]
:+1:

**Comment by bcardiff** [comment]
I would add a link to something like http://davidwalsh.name/squash-commits-git so people can see beforehand what is about. I just google that reference, anyone you like should be fine or, if you prefer, an asciicast in crystal repo itself.

I am not sure if the first entry for example is the oldest or the youngest, clarifying that would raise fewer questions in the reader's mind.

**Comment by asterite** [comment]
:+1:

**Comment by vyp** [comment]
I think the first entry is the oldest iirc.

I think it's fine giving links (maybe to something official?). But at the same time, with external links, you never really have a guarantee of it always existing or being up to date, whereas being in the Contributing.md, this can be controlled by us.

**Comment by jhass** [comment]
Added a clarification about the ordering. I would abstain from an external guide, this way we have something we can update if we change our branching model or similar. Also most guides I've seen rebase ontop of a local master branch, which is often messed up or outdated.

**Comment by adlerhsieh** [comment]
I think the point is on the importance of keeping the commits clean, so I suggest highlight the purpose in the first paragraph, and illustrate details in the following two separate paragraphs. Something like:

===========

### Maintain clean pull requests

All pull requests to this repository are encouraged to include clean, purposeful, and readable commits. It helps us understand your commits as well as pull requests in order to respond more efficiently.

#### Good commits

The commit history should consist of commits that:

- Transform the codebase from one state into another one.
- Motivated by something that should change.
- A bugfix.
- A new feature or some ground work to support a new feature, like changing an existing API or introducing a new isolated class that is later used in the same pull request. 

For example:
```
- Add Enumerable#max? and Enumerable#min?
- Benchmark::IPS bugfix and better comparison alignment
- Removed the anti-feature that allowed invoking methods on a union class. Fix #1061
```

#### Bad commits

The following commit types are strongly discouraged:

- Only showing development history
- Review history

For example:
```
- Start work on X
- More work on X
- Finish X
- Fix comment A
- Fix comment B
```

#### Squashing commits

Remember that review fixes should be squashed into the commits that introduced them. If your change fits well into a single commit, simply keep editing it with `git commit --amend`. Partial staging and committing with `git add -p` and `git commit -p` respectively are also very useful. Another good tool is git stash to put changes aside while switching to another commit. But Git's most useful tool towards this goal is the interactive rebase.

===========

I take your paragraphs and slightly rephrase and re-format. The point is to put emphasis more on the importance of clean commits. This comment is a bit long, hope I make it clear enough :)

**Comment by jhass** [comment]
I'm afraid it's already quite long for my taste, so I'd like to keep it a bit denser.

**Comment by adlerhsieh** [comment]
Yeap definitely, it's just a matter of style.

**Comment by veelenga** [comment]
Contributing guide becomes a longer and longer. Not sure that every newcomer will read it from the beginning to the end. Moreover, this is a general git-related information and should be as concise as possible here. I think it would be better to have it on the wiki in "Interactive rebase and clean commit history" (or something like that) article  and just to link it within Contribution guide. What do you think ?

**Comment by jhass** [comment]
I think with the subsection it's easy to skip intentionally but also harder to miss.

File: Contributing.md [hunk 1]
```diff
@@ -67,6 +67,49 @@ or get from homebrew has a bug (uninstall the LLVM from homebrew too).
 Next, executing `make clean crystal spec` should compile a compiler and using that compiler compile and execute
 the specs. All specs should pass.
 
+## Maintain clean pull requests
+
+The commit history should consist of commits that transform the codebase from one state into another one, motivated by something that
```
Comment (kirbyfan64) [inline Contributing.md]: Uh...isn't this sentence kind of long?

File: Contributing.md [hunk 2]
```diff
@@ -67,6 +67,49 @@ or get from homebrew has a bug (uninstall the LLVM from homebrew too).
...
+should change, be it a bugfix, a new feature or some ground work to support a new feature, like changing an existing API or introducing
+a new isolated class that is later used in the same pull request. It should not show development history ("Start work on X",
+"More work on X", "Finish X") nor review history ("Fix comment A", "Fix comment B"). Review fixes should be squashed into the commits
+that introduced them. If your change fits well into a single commit, simply keep editing it with `git commit --amend`. Partial staging and
+committing with `git add -p` and `git commit -p` respectively are also very useful. Another good tool is `git stash` to put changes aside while
+switching to another commit. But Git's most useful tool towards this goal is the interactive rebase. 
+
+### Doing an interactive rebase
+
+First let's make sure we got a clean reference to rebase upon:
```
Comment (kirbyfan64) [inline Contributing.md]: First**,** let's make sure we **have**

File: Contributing.md [hunk 3]
```diff
@@ -67,6 +67,49 @@ or get from homebrew has a bug (uninstall the LLVM from homebrew too).
...
+
+```sh
+git remote add upstream https://github.com/manastech/crystal.git
+```
+
+That only needs to be done once per clone. Next let's fetch the latest state and start the rebase
```
Comment (kirbyfan64) [inline Contributing.md]: Next**,**

File: Contributing.md [hunk 4]
```diff
@@ -67,6 +67,49 @@ or get from homebrew has a bug (uninstall the LLVM from homebrew too).
...
+
+```sh
+git fetch upstream
+git checkout my_feature_branch
+git rebase -i upstream/master # Or upstream/gh-pages for contributing to the out of code documentation
+```
+
+Now you should be presented with a list of commits in your editor, the first commit you made on your branch at the top. Always keep the first
```
Comment (kirbyfan64) [inline Contributing.md]: "editor, **with** the first commit"?

File: Contributing.md [hunk 5]
```diff
@@ -67,6 +67,49 @@ or get from homebrew has a bug (uninstall the LLVM from homebrew too).
...
+entry at `pick`, however you can reorder the entries. `squash` and `fix` will combine a commit into the one above it, `edit` will pause the
```
Comment (kirbyfan64) [inline Contributing.md]: `pick`**;** however**,** you can

File: Contributing.md [hunk 6]
```diff
@@ -67,6 +67,49 @@ or get from homebrew has a bug (uninstall the LLVM from homebrew too).
...
+rebase so you can edit the commit with `git commit --amend`. In case of conflicts `git mergetool` can be useful to resolve them. To resume a
+paused rebase, either because of a conflict or `edit`, use `git rebase --continue`. Don't worry, you can at any point use `git rebase --abort`
+to return to where you were before the rebase and start from scratch.
+
+Other useful flags to `git commit` are `--fixup` and `--squash`, combined with `git rebase -i --autosquash upstream/master`. Those will create commits that
+are then automatically reordered and marked with `fix` or `squash` respectively.
+
+Once you got a clean history, you can update an existing pull request simply by force pushing to the branch you opened it from. Force pushing is necessary
```
Comment (kirbyfan64) [inline Contributing.md]: Once you **have**

File: Contributing.md [hunk 7]
```diff
@@ -67,6 +67,49 @@ or get from homebrew has a bug (uninstall the LLVM from homebrew too).
...
+since a rebase rewrites history, effectively creating new commits with the same changes. However never do that in the main integration branches (`master`) of
```
Comment (kirbyfan64) [inline Contributing.md]: However**,**

File: Contributing.md [hunk 8]
```diff
@@ -67,6 +67,49 @@ or get from homebrew has a bug (uninstall the LLVM from homebrew too).
...
+your own projects, a not so clean history is to prefer once a commit landed there. Assuming `origin` is your fork on Github, simply:
```
Comment (kirbyfan64) [inline Contributing.md]: **;** a not so clean

File: Contributing.md (hunk 1 above)
Comment (jhass) [inline Contributing.md]: Maybe, though I'm currently rather uncreative at rephrasing it.

**Comment by asterite** [comment]
@jhass I'll soon merge the `formatter` branch into master. Well, I'll probably rebase it. I honestly don't know how to do interactive rebase, squash, etc., so I'll read your guide and see if I understand everything and if there's something missing. So from my side it's a big :+1 for this PR

**Comment by ozra** [comment]
I too lack some git know how on that. So it's welcome. But the tl;dr-factor is important; perhaps footnotes could be used, referring to the detailed technical explanations, placed at the end of the document?

**Pull request merged**